core/src/main/java/org/jboss/jandex/Indexer.java

```diff
@@ -2110,11 +2110,54 @@ private String decodeUtf8Entry(int index) throws IOException {
 
         pos++;
 
-        // DataInputStream needs to read the length again
         int len = (pool[pos] & 0xFF) << 8 | (pool[pos + 1] & 0xFF);
+        final String asciiString = tryDecodeAsciiEntry(pool, pos + 2, len);
+        if (asciiString != null) {
+            return asciiString;
+        }
+        // slow-path: TO IMPROVE!
+        // DataInputStream needs to read the length again
         return new DataInputStream(new ByteArrayInputStream(pool, pos, len + 2)).readUTF();
     }
 
+    private static String tryDecodeAsciiEntry(byte[] chars, int pos, int len) {
```

Comment:
this has been benchmarked in https://github.com/undertow-io/undertow/pull/1424 vs a simple copy loop vs different JDK versions